src/components/ActionButton/StatefulActionButton.ts

```
@@ -0,0 +1,139 @@
+            console.warn(`${state.loggingName ?? 'This state'} is not allowed on this StatefulActionButton.`);
+            return;
+        }
+        this.currentState.onStateExit?.apply(this);
+        state.onStateEntry?.apply(this);
+        this.innerActionButton.updateIcon(state.icon);
+        this.innerActionButton.updateTooltip(state.tooltip);
+        this.currentState = state;
+    }
+
+    /**
+     * Return the current state of the instance.
+     */
+    public getCurrentState() {
+        return this.currentState;
+    }
+
+    /**
+     * Check if the options given to the constructor are valid.
+     * If not, it will also display the appropriate warning.
+     */
+    private areOptionsValid(): boolean {
+        if (!this.options.states?.length) {
+            console.warn('The stateful action button cannot render if no states are defined.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        if (!this.options.initalState) {
+            console.warn('The stateful action button cannot render if the inital states is not defined.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        if (this.options.states.indexOf(this.options.initalState) < 0) {
+            console.warn('The stateful action button cannot render if the initial state is not in the list of states.');
+            Coveo.$$(this.element).hide();
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Check if a transition from the current state to @param state is allowed.
+     * @param state the destination of the transition
+     */
+    private isTransitionAllowed(state: IStatefulActionButtonState) {
+        return (
+            this.options.states.indexOf(state) > -1 &&
+            (!this.options.allowedTransitions ||
+                this.options.allowedTransitions.some((transition) => transition.from === this.currentState && transition.to === state))
```

As discussed offline, we can continue to compare the states by reference. But we should add more validations to ensure that `states` and `allowedTransitions` are using the exact same references. Also, it should raise an error if the implementer passes an unknown state reference to the `switchTo` method.

---

👍 Should be good now ;)